### DISCUSSION TITLE
onRecordsListRequest - How can I get the user who listed

### DISCUSSION THREAD

**Discussion by emiliobiasi** [pocketbase/pocketbase#5099] [Q&A] [answered]
I'm attempting to retrieve some data from the user who was logged in and listed the records from my "Posts" collection.
I want the user id to have a posts_views_control collection and increment the post view only in some conditions.
Am I missing something? Everytime I try to get any auth information, my collection Posts turns to an empty collection.

My code increments +1 every time some user list the records, but Im trying to increment +1 only if the user never saw the post.

```js
onRecordsListRequest((e) => {
  console.log(JSON.stringify(e.httpContext));
  console.log(JSON.stringify(e.result));
  console.log(JSON.stringify(e));
  const records = e.records;
  console.log("records: ", JSON.stringify(records));
  records.forEach((record) => {
    $app.dao().runInTransaction((txDao) => {
      const n = txDao.findRecordById(record.collection().id, record.id);
      n.set("views", record.get("views") + 1);
      txDao.saveRecord(n);
    });
  });
}, "posts");
```

**Comment by ganigeorgiev** [comment, marked as answer]
> Everytime I try to get any auth information, my collection Posts turns to an empty collection.

I'm not sure that I understand what do you mean by this. If you mean that the above `console.log` statements are empty then it is possible because they are not native JS objects and may not have public serializable fields or safe string representation to show.

If you want to access the current auth state you can check ["Retrieving the current auth state"](https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state). Note that the user could be either an admin or regular auth record user, so for example you may have to skip the counter if it is admin (_this is simplified with the refactoring and there is only 1 auth context because the admins are converted to a special system auth collection_).


>  Im trying to increment +1 only if the user never saw the post

I'm not familiar with your data structure but note that if you are currently storing the counter as part of the listed record then you'll need to reorganize your data to keep track of the `userId-postId` pairs. Or in other words, if you just need a "seen/unseen" flag you could:

1. Create a helper collection, eg. `userPostSeen`, with the following fields:
   - `user` - single _nonempty_ `relation` field to the `users` (enable cascade to delete automatically on users delete)
   - `post` - single _nonempty_ `relation` field to the `posts` (enable cascade to delete automatically on posts delete)
   
   1.1. Define a combined `UNIQUE` index constraint for the `user` and `post` fields to prevent duplicates, eg:
   ```sql
   CREATE UNIQUE INDEX `idx_wF9yttg` ON `userPostSeen` (`user`, `posts`)
   ```

   1.2. To allow the individual user to query their "seen" status you can also add as List and View API rule to the collection something like:
   ```js
   @request.auth.id != "" && @request.auth.id = user
   ```

2. To perform the update from your list hook you could do something like (_I haven't tested it_):
    ```js
    onRecordsListRequest((e) => {
        const info = $apis.requestInfo(c);

        if (info.admin || !info.authRecord) {
            return // skip for admins or unauthorized records (the latter depends on your List API rule)
        }

        const userPostSeenCollection = $app.dao().findCollectionByNameOrId("userPostSeen")

        for (let record of e.records) {
            try {
                // throws if missing
                app.dao().findFirstRecordByFilter(userPostSeenCollection.id, "user={:user} && post={:post}", {
                    user: info.authRecord.id,
                    post: record.id,
                })
            } catch (_) {
                try {
                    // create
                    const seenRel = new Record()
                    seenRel.set("user", info.authRecord.id)
                    seenRel.set("post", record.id)
                } catch (err) {
                    // if the create fails it is most likely due to a concurrnet request hitting the unique constraint
                    // we can simply ignore it and log it if needed
                    $app.logger().warn("failed to save user post seen state", "error", err.toString())
                }
            }
        }
    }, "posts");
    ```

3. From the client-side, if you want to fetch the individual user seen status for a specific post you can either send an individual `getFirstListItem` request to the `userPostSeen` collection. Alternatively you can also retrieve the user relations using [back-relation expand](https://pocketbase.io/docs/working-with-relations/#back-relations):

   ```js
   const posts = await pb.collection("posts").getFullList({
     expand: "userPostSeen_via_post"
   })
   
   // to check whether a single post item is "seen" by the user the expanded relation must be empty, ex.:
   // const isSeen = !!posts[0].expand?.userPostSeen_via_post?.length
   ```

Note that the above is just one way to do it. Depending on your use case an alternative could be also to have a multiple `relation` field in the `users` collection pointing to `posts` that will contain entries of only the "seen" posts.

**Reply by ganigeorgiev** [reply]
If you need to get a counter for how many users has seen a post, you could create a helper View collection, ex. `postsSeenCount`, with the following query:

```sql
-- alterntatively if you want the view to include also posts that has never been viewed you can select from "posts" and join the userPostSeen with it in a similar manner.
SELECT
   ups.post as id,
   ups.post as post, -- will be used as relation field
   count(*) as total
FROM userPostSeen as ups
GROUP BY ups.post
```
If you want everyone to be able to access the views count update the List and View API rule to be empty (or only allow for registered users, ex. `@request.auth.id != ""`).

Then on the client-side you can use again [back-relation expand](https://pocketbase.io/docs/working-with-relations/#back-relations) to get the views count together with your posts:

```js
const posts = await pb.collection("posts").getFullList({
  expand: "userPostSeen_via_post, postsSeenCount_via_post"
})
```
